### PR TITLE
Navigate substructure even for different kinds on ASTRule

### LINKED ISSUE #1237: unused_optional_binding misses ifs inside functions
This example does not trigger the unused_optional_binding:

```swift
func foo() {
    if let _ = bar {
    }
}
```

If you add this test case:

```swift
"func foo() {\nif let _ = bar {\n}\n"
```

You can see that the tests fail to pass. I believe this is because the `if` in this case is nested in the substructure of the func, which isn't handled by the [current logic](https://github.com/realm/SwiftLint/blob/master/Source/SwiftLintFramework/Rules/UnusedOptionalBindingRule.swift#L50)

// @rakaramos

### PR DESCRIPTION
**Pull request by marcelofabri**
Fixes #1237

I expect oss-check to introduce some violations that weren't being caught before.

### CODE REVIEW

File: Source/SwiftLintFramework/Rules/UnusedOptionalBindingRule.swift [hunk 1]
```diff
@@ -40,23 +42,31 @@ public struct UnusedOptionalBindingRule: ASTRule, ConfigurationProviderRule {
...
 
     public func validate(file: File,
                          kind: StatementKind,
                          dictionary: [String: SourceKitRepresentable]) -> [StyleViolation] {
+        let conditionKind = "source.lang.swift.structure.elem.condition_expr"
         guard kind == .if || kind == .guard,
-            let offset = dictionary.offset,
-            let length = dictionary.length,
-            let range = file.contents.bridge().byteRangeToNSRange(start: offset, length: length) else {
+            let elements = dictionary.elements?.filter({ $0.kind == conditionKind }) else {
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/UnusedOptionalBindingRule.swift]: this was needed to make sure that the pattern was in the condition, not inside an `if` body

**Comment by keith** [comment]
Thanks for submitting this one! I'll test this locally shortly, FWIW the `RedundantNilCoalescingRule` also has the non-recursive issue on master as well.

**Comment by marcelofabri** [comment]
@keith I think that might be another issue because that rule isn't an `ASTRule`

**Comment by keith** [comment]
Sorry, wrong rule, I meant the `RedundantOptionalInitializationRule` which is 😄

**Comment by keith** [comment]
It looks like with this test case in `RedundantOptionalInitializationRule` :

```
            "func foo() {\nvar myVar: Optional<Int>↓=nil\n}\n"
```

This still fails on this branch

**Comment by keith** [comment]
Ah, it looks like there's some duplicate custom logic in that rule that we'll need to fix as well.

**Comment by marcelofabri** [comment]
> Ah, it looks like there's some duplicate custom logic in that rule that we'll need to fix as well.

From a quick look I haven't found a motive to fail. My bet is that it's the same problem as #136.

**Comment by marcelofabri** [comment]
In fact, running `sourcekitten structure`:


```json
{
  "key.diagnostic_stage" : "source.diagnostic.stage.swift.parse",
  "key.substructure" : [
    {
      "key.offset" : 0,
      "key.nameoffset" : 5,
      "key.accessibility" : "source.lang.swift.accessibility.internal",
      "key.length" : 47,
      "key.name" : "foo()",
      "key.bodyoffset" : 12,
      "key.kind" : "source.lang.swift.decl.function.free",
      "key.namelength" : 5,
      "key.bodylength" : 34
    }
  ],
  "key.offset" : 0,
  "key.length" : 48
}
```

**Comment by keith** [comment]
Nope, I think this fixes what it meant to. This just didn't fix what I was hoping, but that's not the fault of this code, just my misunderstanding!

**Comment by jpsim** [comment]
🚢 ❗️

**Pull request merged**